// graphile-worker/dist/getTasks.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getTasksInternal = exports.getTasks = void 0;
const promises_1 = require("fs/promises");
const path_1 = require("path");
const fs_1 = require("./fs");
const interfaces_1 = require("./interfaces");
const lib_1 = require("./lib");
const DIRECTORY_REGEXP = /^[A-Za-z0-9_-]+$/;
const FILE_REGEXP = /^([A-Za-z0-9_-]+)((?:\.[A-Za-z0-9_-]+)*)$/;
function validTasks(logger, obj) {
    const tasks = {};
    Object.keys(obj).forEach((taskIdentifier) => {
        const task = obj[taskIdentifier];
        if ((0, interfaces_1.isValidTask)(task)) {
            tasks[taskIdentifier] = task;
        }
        else {
            logger.warn(`Not a valid task '${taskIdentifier}' - expected function, received ${task ? typeof task : String(task)}.`, {
                invalidTask: true,
                task,
                taskIdentifier,
            });
        }
    });
    return tasks;
}
async function loadFileIntoTasks(logger, tasks, filename, name = null) {
    const rawMod = await import(filename);
    // Normally, import() of a commonJS module with `module.exports` write
    // would result in `{ default: module.exports }`.
    // TypeScript in CommonJS mode when imported with Node ESM can lead to
    // two levels of `__esModule: true`; so we try and grab the inner one
    // if we can.
    const mod = rawMod.default?.default?.__esModule === true
        ? rawMod.default.default
        : rawMod.default?.__esModule === true
            ? rawMod.default
            : Object.keys(rawMod).length === 1 &&
                typeof rawMod.default === "object" &&
                rawMod.default !== null
                ? rawMod.default
                : rawMod;
    if (name) {
        // Always take the default export if there is one
        const task = mod.default || mod;
        if ((0, interfaces_1.isValidTask)(task)) {
            tasks[name] = task;
        }
        else {
            throw new Error(`Invalid task '${name}' - expected function, received ${task ? typeof task : String(task)}.`);
        }
    }
    else {
        Object.keys(tasks).forEach((taskIdentifier) => {
            delete tasks[taskIdentifier];
        });
        if (!mod.default || typeof mod.default === "function") {
            Object.assign(tasks, validTasks(logger, mod));
        }
        else {
            Object.assign(tasks, validTasks(logger, mod.default));
        }
    }
}
async function getTasks(options, taskPath) {
    const compiledSharedOptions = (0, lib_1.processSharedOptions)(options);
    const result = await getTasksInternal(compiledSharedOptions, taskPath);
    // This assign is used in `__tests__/getTasks.test.ts`
    return Object.assign(result, { compiledSharedOptions });
}
exports.getTasks = getTasks;
async function getTasksInternal(compiledSharedOptions, taskPath) {
    const { logger } = compiledSharedOptions;
    const pathStat = await (0, fs_1.tryStat)(taskPath);
    if (!pathStat) {
        throw new Error(`Could not find tasks to execute - taskDirectory '${taskPath}' does not exist`);
    }
    const tasks = Object.create(null);
    if (pathStat.isFile()) {
        // Try and require it
        await loadFileIntoTasks(logger, tasks, taskPath, null);
    }
    else if (pathStat.isDirectory()) {
        const collectedTaskPaths = Object.create(null);
        await getTasksFromDirectory(compiledSharedOptions, collectedTaskPaths, taskPath, []);
        const taskIdentifiers = Object.keys(collectedTaskPaths).sort((a, z) => a.localeCompare(z, "en-US"));
        for (const taskIdentifier of taskIdentifiers) {
            const fileDetailsList = collectedTaskPaths[taskIdentifier];
            const event = {
                handler: undefined,
                taskIdentifier,
                fileDetailsList,
            };
            await compiledSharedOptions.hooks.process("loadTaskFromFiles", event);
            const handler = event.handler;
            if (handler) {
                tasks[taskIdentifier] = handler;
            }
            else {
                logger.warn(`Failed to load task '${taskIdentifier}' - no supported handlers found for path${fileDetailsList.length > 1 ? "s" : ""}: '${fileDetailsList.map((d) => d.fullPath).join("', '")}'`);
            }
        }
    }
    let released = false;
    return {
        tasks,
        compiledSharedOptions,
        release: () => {
            if (released) {
                return;
            }
            released = true;
        },
    };
}
exports.getTasksInternal = getTasksInternal;
async function getTasksFromDirectory(compiledSharedOptions, collectedTaskPaths, taskPath, subpath) {
    const { logger } = compiledSharedOptions;
    const folderPath = (0, path_1.join)(taskPath, ...subpath);
    // Try and require its contents
    const entries = await (0, promises_1.readdir)(folderPath);
    await Promise.all(entries.map(async (entry) => {
        const fullPath = (0, path_1.join)(taskPath, ...subpath, entry);
        const stats = await (0, promises_1.lstat)(fullPath);
        if (stats.isDirectory()) {
            if (DIRECTORY_REGEXP.test(entry)) {
                await getTasksFromDirectory(compiledSharedOptions, collectedTaskPaths, taskPath, [...subpath, entry]);
            }
            else {
                logger.info(`Ignoring directory '${fullPath}' - '${entry}' does not match allowed regexp.`);
            }
        }
        else if (stats.isSymbolicLink()) {
            // Must be a symbolic link to a file, otherwise ignore
            const symlinkTarget = await (0, promises_1.realpath)(fullPath);
            const targetStats = await (0, promises_1.lstat)(symlinkTarget);
            if (targetStats.isFile() && !targetStats.isSymbolicLink()) {
                maybeAddFile(compiledSharedOptions, collectedTaskPaths, subpath, entry, symlinkTarget, targetStats);
            }
        }
        else if (stats.isFile()) {
            maybeAddFile(compiledSharedOptions, collectedTaskPaths, subpath, entry, fullPath, stats);
        }
    }));
}
function maybeAddFile(compiledSharedOptions, collectedTaskPaths, subpath, entry, fullPath, stats) {
    const { logger } = compiledSharedOptions;
    const matches = FILE_REGEXP.exec(entry);
    if (matches) {
        const [, baseName, extension] = matches;
        const entry = {
            fullPath,
            stats,
            baseName,
            extension,
        };
        const taskIdentifier = [...subpath, baseName].join("/");
        if (!collectedTaskPaths[taskIdentifier]) {
            collectedTaskPaths[taskIdentifier] = [entry];
        }
        else {
            collectedTaskPaths[taskIdentifier].push(entry);
        }
    }
    else {
        logger.info(`Ignoring file '${fullPath}' - '${entry}' does not match allowed regexp.`);
    }
}
